tawazi/dag.py

```diff
@@ -48,89 +56,49 @@ def dfs(n: ExecNode, graph: DiGraphEx, visited: Set[ExecNode]) -> None:
-        self.priority = priority
-        self.is_sequential = is_sequential
-        self.logger = logger
+        # works by pruning the graph until all leaf nodes
+        # are contained inside the provided "nodes"
+        # in the arguments of this method
 
-        # a string that identifies the ExecNode.
-        # It is either the name of the identifying function or the identifying string id_
-        self.__name__ = self.exec_function.__name__ if not isinstance(id_, str) else id_
-
-        # Attempt to automatically assign a good enough argument name
-        if isinstance(argument_name, str) and argument_name != "":
-            self.argument_name = argument_name
-        else:
-            self.argument_name = (
-                self.id.__name__ if isinstance(self.id, FunctionType) else str(self.id)
+        if any([node not in self.nodes for node in nodes]):
+            raise ValueError(
+                f"The provided nodes are not in the graph. "
+                f"The provided nodes are: {nodes}."
+                f"The graph only contain: {self.nodes}."
             )
 
-        # todo remove and make ExecNode immutable
-        self.result: Optional[Dict[str, Any]] = None
+        leaf_nodes = self.leaf_nodes()
+        nodes_to_remove = set(leaf_nodes).difference(set(nodes))
 
-    def __repr__(self) -> str:
-        return f"{self.__class__.__name__} {self.id} ~ | <{hex(id(self))}>"
-
-    @property
-    def computed_dependencies(self) -> bool:
-        return isinstance(self.depends_on, list)
-
-    # this is breaking change however
-    def execute(self, node_dict: Dict[Hashable, "ExecNode"]) -> Optional[Dict[str, Any]]:
-        """
-        Execute the ExecNode directly or according to an execution graph.
-        Args:
-            node_dict (Dict[Hashable, ExecNode]): A shared dictionary containing the other ExecNodes in the DAG;
-                                                the key is the id of the ExecNode.
+        while nodes_to_remove:
+            node_to_remove = nodes_to_remove.pop()
+            self.remove_node(node_to_remove)
 
-        Returns: the result of the execution of the current ExecNode
-        """
-        # 1. fabricate the arguments for this ExecNode
-        self.logger.debug(f"Start executing {self.id} with task {self.exec_function}")
+            leaf_nodes = self.leaf_nodes()
+            nodes_to_remove = set(leaf_nodes).difference(set(nodes))
 
-        kwargs = {
-            node_dict[dep_hash].argument_name: node_dict[dep_hash].result
-            for dep_hash in self.depends_on
-        }
+        leaf_nodes = self.leaf_nodes()
+        impossible_to_remove_nodes = set(nodes).difference(set(leaf_nodes))
 
-        # 2. write the result
-        self.result = self.exec_function(**kwargs)
+        if len(impossible_to_remove_nodes) > 0:
+            logger_.warning(
+                f"The provided nodes contain more nodes than necessary, "
```

Comment:
no... in this case, the problem comes from the fact that the user provided more nodes than necessary. for example:
a graph of the following shape:
```
a
|  \
b  c
```

Now imagine the user provided the nodes `a` and `b` to be ran... The code will run `a` then `b` and avoid running `c`
Although this runs without a problem... it shouldn't be used this way, instead the user must provide only the leaf nodes which he wants to be executed because he has interest in its values... 

But now when I think about it... maybe the user is also interested in the values generated by `a` and he doesn't care about the relationship between `a` and `b`... so I shouldn't actually raise a warning at the end

Reply:
aaa right I see